File: backend/src/aerospike_py_admin_ui_api/models/query.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from .record import AerospikeRecord, BinValue


class QueryPredicate(BaseModel):
    bin: str = Field(min_length=1)
    operator: Literal["equals", "between", "contains", "geo_within_region", "geo_contains_point"]
    value: BinValue
    value2: BinValue | None = None


class QueryRequest(BaseModel):
    namespace: str = Field(min_length=1)
    set: str | None = None
    predicate: QueryPredicate | None = None
    selectBins: list[str] | None = None
    expression: str | None = None
    maxRecords: int | None = Field(default=None, ge=1, le=1_000_000)
    primaryKey: str | None = None


class QueryResponse(BaseModel):
    records: list[AerospikeRecord]
    executionTimeMs: int = Field(ge=0)
    scannedRecords: int = Field(ge=0)
    returnedRecords: int = Field(ge=0)

File: frontend/src/lib/validations/query.ts
import { z } from "zod";

const binValueSchema: z.ZodType = z.lazy(() =>
  z.union([
    z.string(),
    z.number(),
    z.boolean(),
    z.null(),
    z.array(binValueSchema),
    z.record(z.string(), binValueSchema),
  ]),
);

export const queryPredicateSchema = z.object({
  bin: z.string().min(1, "Bin name is required"),
  operator: z.enum(["equals", "between", "contains", "geo_within_region", "geo_contains_point"], {
    message: "Operator is required",
  }),
  value: binValueSchema,
  value2: binValueSchema.optional(),
});

export const queryRequestSchema = z.object({
  namespace: z.string().min(1, "Namespace is required"),
  set: z.string().optional(),
  predicate: queryPredicateSchema.optional(),
  selectBins: z.array(z.string()).optional(),
  expression: z.string().optional(),
  maxRecords: z.number().int().positive().optional(),
  primaryKey: z.string().optional(),
});

export type QueryPredicateFormValues = z.infer<typeof queryPredicateSchema>;
export type QueryRequestFormValues = z.infer<typeof queryRequestSchema>;
